src/Microsoft.Data/ArrowStringColumn.cs

```diff
@@ -0,0 +1,348 @@
+                {
+                    mutableDataBuffer = new MutableDataFrameBuffer<byte>();
+                    _dataBuffers.Add(mutableDataBuffer);
+                    _nullBitMapBuffers.Add(new MutableDataFrameBuffer<byte>());
+                    var offsetBuffer = new MutableDataFrameBuffer<int>();
+                    _offsetsBuffers.Add(offsetBuffer);
+                    offsetBuffer.Append(0);
+                }
+                mutableDataBuffer.EnsureCapacity(bytes.Length);
+                bytes.AsMemory().CopyTo(mutableDataBuffer.Memory.Slice(mutableDataBuffer.Length));
+                mutableDataBuffer.Length += bytes.Length;
+                mutableOffsetsBuffer.Append(mutableOffsetsBuffer[mutableOffsetsBuffer.Length - 1] + bytes.Length);
+            }
+            SetValidityBit(Length - 1, value == null ? true : false);
+        }
+
+        private int GetBufferIndexContainingRowIndex(ref long rowIndex)
+        {
+            if (rowIndex >= Length)
+            {
+                throw new ArgumentOutOfRangeException(Strings.ColumnIndexOutOfRange, nameof(rowIndex));
+            }
+
+            // Since the strings here could be of variable length, scan linearly
+            int curArrayIndex = 0;
+            int numBuffers = _offsetsBuffers.Count;
+            while (curArrayIndex < numBuffers && rowIndex > _offsetsBuffers[curArrayIndex].Length - 1)
+            {
+                rowIndex -= _offsetsBuffers[curArrayIndex].Length - 1;
+                curArrayIndex++;
+            }
+            return curArrayIndex;
+        }
+
+        private ReadOnlySpan<byte> GetBytes(long index)
+        {
+            int offsetsBufferIndex = GetBufferIndexContainingRowIndex(ref index);
+            int currentOffset = _offsetsBuffers[offsetsBufferIndex].ReadOnlySpan[(int)index];
+            int nextOffset = _offsetsBuffers[offsetsBufferIndex].ReadOnlySpan[(int)index + 1];
+            int numberOfBytes = nextOffset - currentOffset;
+            return _dataBuffers[offsetsBufferIndex].ReadOnlySpan.Slice(currentOffset, numberOfBytes);
+        }
+
+        protected override object GetValue(long rowIndex)
+        {
+            var bytes = GetBytes(rowIndex);
+            unsafe
+            {
+                fixed (byte* data = &MemoryMarshal.GetReference(bytes))
+                    return _encoding.GetString(data, bytes.Length);
+            }
+        }
+
+        protected override object GetValue(long startIndex, int length)
+        {
+            var ret = new List<string>();
+            while (ret.Count < length)
+            {
+                ret.Add((string)GetValue((int)startIndex));
+            }
+            return ret;
+        }
+
+        protected override void SetValue(long rowIndex, object value)
+        {
+            throw new NotSupportedException(Strings.ImmutableStringColumn);
+        }
+
+        public new string this[long rowIndex]
+        {
+            get => (string)GetValue(rowIndex);
```

Comment:
This is backwards. The weakly typed API (the one returning `object`) should call the strongly-typed API (the one returning a `string`). Then no casts are necessary.

Reply:
Both `GetValue` and `new string this[long rowIndex]` are overriding weakly typed methods in `BaseColumn`. If I made `GetValue` call `this[rowIndex]`, it *may* work, but it'd rather not rely on method hiding like that. I did make a new `GetValueImplementation` method though that is strongly typed, so there are no casts now.